File: agents/example_java_agent/src/Agent.java
import java.io.*;
import java.net.*;
import java.nio.*;
import java.util.*;

import hex.*;

public class Agent {
	private static Net net;
	private static State state;
	private static Board board;
	private static HexPlayer player, opponent, winner;
	private static int gameSecs, threadLimit, memLimitMib; // unused

	public static void main(String[] args) {
		if (args.length < 2) {
			System.err.println("Not enough args: <host> <port>");
			return;
		}

		String host = args[0];
		int port = Integer.parseInt(args[1]);

		try {
			net = new Net(host, port);
		} catch (IOException ex) {
			System.err.println("Failed to initialise network");
			return;
		}

		state = State.START;

		boolean gameOver = false, firstRound = true;
		while (!gameOver) {
			switch (state) {
			case START -> {
				Optional<NetMessage> msg = net.recvMsg();

				if (!msg.isPresent()) {
					System.err.println("Failed to receive message from hex server");
					return;
				}

				if (msg.get() instanceof StartMessage start) {
					player = start.player();

					gameSecs = start.gameSecs();
					threadLimit = start.threadLimit();
					memLimitMib = start.memLimitMib();

					board = new Board(start.boardSize());

					switch (player) {
					case BLACK -> { opponent = HexPlayer.WHITE; state = State.SEND; }
					case WHITE -> { opponent = HexPlayer.BLACK; state = State.RECV; }
					}
				} else {
					System.err.println("Invalid message received from server");
					return;
				}
			}

			case RECV -> {
				Optional<NetMessage> msg = net.recvMsg();

				if (!msg.isPresent()) {
					System.err.println("Failed to receive message from hex server");
					return;
				}

				if (msg.get() instanceof MoveMessage move) {
					board.play(opponent, move.boardX(), move.boardY());

					if (firstRound && new Random().nextBoolean()) {
						board.swap();

						if (!net.sendMsg(new SwapMessage())) {
							System.err.println("Failed to send message to hex server");
							return;
						}

						state = state.RECV;
					} else {
						state = state.SEND;
					}
				} else if (msg.get() instanceof SwapMessage swap) {
					board.swap();
					state = State.SEND;
				} else if (msg.get() instanceof EndMessage end) {
					winner = end.winner();
					state = State.END;
				} else {
					System.err.println("Invalid message received from server");
					return;
				}

				firstRound = false;
			}

			case SEND -> {
				Optional<Move> nextMove = board.next();

				if (!nextMove.isPresent()) {
					System.err.println("Failed to generate next board move");
					return;
				}

				Move move = nextMove.get();

				board.play(player, move.x(), move.y());

				if (!net.sendMsg(new MoveMessage(move.x(), move.y()))) {
					System.err.println("Failed to send message to hex server");
					return;
				}

				state = State.RECV;
				firstRound = false;
			}

			case END -> {
				gameOver = true;
			}
			}
		}

		return;
	}
}

enum State {
	START, RECV, SEND, END,
}

record Move(int x, int y) {}

enum Cell {
	BLACK(HexPlayer.BLACK.value),
	WHITE(HexPlayer.WHITE.value),
	EMPTY(2);

	public final int value;

	private Cell(int value) {
		this.value = value;
	}
}

class Board {
	private final int size;
	private final ArrayList<Cell> cells;
	private final ArrayList<Move> moves;

	public Board(int size) {
		this.size = size;
		this.cells = new ArrayList<>(size * size);
		this.moves = new ArrayList<>(size * size);

		for (int j = 0; j < size; j++) {
			for (int i = 0; i < size; i++) {
				this.cells.add(Cell.EMPTY);
				this.moves.add(new Move(i, j));
			}
		}

		Collections.shuffle(this.moves);
	}

	public boolean play(HexPlayer player, int x, int y) {
		int idx = y * this.size + x;

		Cell cell = this.cells.get(idx);
		if (cell != Cell.EMPTY) return false;

		switch (player) {
		case BLACK -> this.cells.set(idx, Cell.BLACK);
		case WHITE -> this.cells.set(idx, Cell.WHITE);
		}

		this.moves.remove(new Move(x, y));

		return true;
	}

	public void swap() {
		this.moves.clear();

		for (int j = 0; j < this.size; j++) {
			for (int i = 0; i < this.size; i++) {
				int idx = j * this.size + i;

				Cell cell = this.cells.get(idx);

				switch (cell) {
				case BLACK -> this.cells.set(idx, Cell.WHITE);
				case WHITE -> this.cells.set(idx, Cell.BLACK);
				case EMPTY -> this.moves.add(new Move(i, j));
				}
			}
		}

		Collections.shuffle(this.moves);
	}

	public Optional<Move> next() {
		if (this.moves.isEmpty()) return Optional.empty();

		return Optional.of(this.moves.remove(this.moves.size() - 1));
	}
}

sealed interface NetMessage {}

record StartMessage(HexPlayer player, int boardSize, int gameSecs, int threadLimit, int memLimitMib) implements NetMessage {}
record MoveMessage(int boardX, int boardY) implements NetMessage {}
record SwapMessage() implements NetMessage {}
record EndMessage(HexPlayer winner) implements NetMessage {}

class Net {
	private final Socket sock;
	private final OutputStream out;
	private final InputStream in;

	public static final int MESSAGE_SIZE = 32;

	public Net(String host, int port) throws IOException {
		this.sock = new Socket(host, port);

		this.out = this.sock.getOutputStream();
		this.in = this.sock.getInputStream();
	}

	public Optional<NetMessage> recvMsg() {
		byte[] buf = new byte[MESSAGE_SIZE];

		int nbytes_recv = 0;

		do {
			int curr;
			try {
				curr = this.in.read(buf, nbytes_recv, buf.length - nbytes_recv);
			} catch (IOException ex) {
				return Optional.empty();
			}

			if (curr <= 0) return Optional.empty();
			nbytes_recv += curr;
		} while (nbytes_recv < buf.length);

		return deserialiseMsg(ByteBuffer.wrap(buf));
	}

	public boolean sendMsg(NetMessage msg) {
		byte[] buf = new byte[MESSAGE_SIZE];

		serialiseMsg(msg, ByteBuffer.wrap(buf));

		try {
			this.out.write(buf);
			this.out.flush();
		} catch (IOException ex) {
			return false;
		}

		return true;
	}

	private static void serialiseMsg(NetMessage msg, ByteBuffer buf) {
		buf.order(ByteOrder.BIG_ENDIAN);

		if (msg instanceof StartMessage start) {
			buf.putInt(HexMessageType.START.value);
			buf.putInt(start.player().value);
			buf.putInt(start.boardSize());
			buf.putInt(start.gameSecs());
			buf.putInt(start.threadLimit());
			buf.putInt(start.memLimitMib());
		} else if (msg instanceof MoveMessage move) {
			buf.putInt(HexMessageType.MOVE.value);
			buf.putInt(move.boardX());
			buf.putInt(move.boardY());
		} else if (msg instanceof SwapMessage swap) {
			buf.putInt(HexMessageType.SWAP.value);
		} else if (msg instanceof EndMessage end) {
			buf.putInt(HexMessageType.END.value);
			buf.putInt(end.winner().value);
		}
	}

	private static Optional<NetMessage> deserialiseMsg(ByteBuffer buf) {
		buf.order(ByteOrder.BIG_ENDIAN);

		int type = buf.getInt();

		if (type == HexMessageType.START.value) {
			HexPlayer player = HexPlayer.fromRaw(buf.getInt());
			int boardSize = buf.getInt();
			int gameSecs = buf.getInt();
			int threadLimit = buf.getInt();
			int memLimitMib = buf.getInt();
			return Optional.of(new StartMessage(player, boardSize, gameSecs, threadLimit, memLimitMib));
		} else if (type == HexMessageType.MOVE.value) {
			int boardX = buf.getInt();
			int boardY = buf.getInt();
			return Optional.of(new MoveMessage(boardX, boardY));
		} else if (type == HexMessageType.SWAP.value) {
			return Optional.of(new SwapMessage());
		} else if (type == HexMessageType.END.value) {
			HexPlayer winner = HexPlayer.fromRaw(buf.getInt());
			return Optional.of(new EndMessage(winner));
		} else {
			return Optional.empty();
		}
	}
}

File: agents/example_cpp_agent/src/agent.cpp
#include "agent.hpp"

enum class Cell {
	BLACK = HEX_PLAYER_BLACK,
	WHITE = HEX_PLAYER_WHITE,
	EMPTY,
};

struct Move {
	u32 x, y;

	bool operator==(const Move &rhs) {
		return this->x == rhs.x && this->y == rhs.y;
	}
};

void swap(Move &lhs, Move &rhs) {
	std::swap(lhs.x, rhs.x);
	std::swap(lhs.y, rhs.y);
}

class Board {
	u32 size;
	std::vector<Cell> cells;
	std::vector<Move> moves;

public:
	template <class URBG>
	Board(u32 size, URBG &&rng) : size(size) {
		cells.reserve(size * size);
		moves.reserve(size * size);

		for (u32 j = 0; j < this->size; j++) {
			for (u32 i = 0; i < this->size; i++) {
				this->cells.push_back(Cell::EMPTY);

				Move move{i, j};
				this->moves.push_back(move);
			}
		}

		std::shuffle(this->moves.begin(), this->moves.end(), rng);
	}

	bool play(enum hex_player player, u32 x, u32 y) {
		Cell &cell = this->cells.at(y * this->size + x);
		if (cell != Cell::EMPTY) return false;

		switch (player) {
		case HEX_PLAYER_BLACK:
			cell = Cell::BLACK;
			break;

		case HEX_PLAYER_WHITE:
			cell = Cell::WHITE;
			break;
		}

		Move move{x, y};
		auto it = std::find(this->moves.begin(), this->moves.end(), move);
		if (it != std::end(this->moves)) {
			::swap(*it, this->moves.back());
			this->moves.pop_back();
		}

		return true;
	}

	template <class URBG>
	void swap(URBG &&rng) {
		this->moves.clear();

		for (u32 j = 0; j < this->size; j++) {
			for (u32 i = 0; i < this->size; i++) {
				Cell &cell = this->cells.at(j * this->size + i);

				switch (cell) {
				case Cell::BLACK: cell = Cell::WHITE; break;
				case Cell::WHITE: cell = Cell::BLACK; break;
				case Cell::EMPTY:
					Move move{i, j};
					this->moves.push_back(move);
					break;
				}
			}
		}

		std::shuffle(this->moves.begin(), this->moves.end(), rng);
	}

	bool next(Move &out) {
		if (this->moves.empty()) return false;

		out = this->moves.back();
		this->moves.pop_back();

		return true;
	}
};

class Net {
	int sockfd;
public:
	Net() : sockfd(-1) {}

	bool init(char *host, char *port) {
		struct addrinfo hints, *addrinfo, *ptr;

		memset(&hints, 0, sizeof hints);
		hints.ai_family = AF_UNSPEC;
		hints.ai_socktype = SOCK_STREAM;

		int res;
		if ((res = getaddrinfo(host, port, &hints, &addrinfo))) {
			std::cerr << "getaddrinfo: " << gai_strerror(res) << std::endl;
			return false;
		}

		int sockfd;
		for (ptr = addrinfo; ptr; ptr = ptr->ai_next) {
			sockfd = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol);
			if (sockfd == -1) continue;
			if (connect(sockfd, ptr->ai_addr, ptr->ai_addrlen) != -1) break;
			close(sockfd);
		}

		freeaddrinfo(addrinfo);

		if (!ptr) {
			std::cerr << "Failed to connect to " << host << ":" << port << std::endl;
			return false;
		}

		this->sockfd = sockfd;

		return true;
	}

	bool recv_msg(struct hex_msg &out, const std::vector<enum hex_msg_type> &expected) {
		u8 buf[HEX_MSG_SZ];

		size_t nbytes_recv = 0, len = HEX_MSG_SZ;

		do {
			ssize_t curr = recv(this->sockfd, buf + nbytes_recv, len - nbytes_recv, 0);
			if (curr <= 0) return false; // error or socket shutdown
			nbytes_recv += curr;
		} while (nbytes_recv < len);

		struct hex_msg msg;
		if (!hex_msg_try_deserialise(buf, &msg)) return false;

		if (std::find(expected.begin(), expected.end(), msg.type) != std::end(expected)) {
			out = msg;
			return true;
		}

		return false;
	}

	bool send_msg(const struct hex_msg &msg) {
		u8 buf[HEX_MSG_SZ];
		if (!hex_msg_try_serialise(&msg, buf)) return false;

		size_t nbytes_sent = 0, len = HEX_MSG_SZ;

		do {
			ssize_t curr = send(this->sockfd, buf + nbytes_sent, len - nbytes_sent, 0);
			if (curr <= 0) return false; // error or socket shutdown
			nbytes_sent += curr;
		} while (nbytes_sent < len);

		return true;
	}
};

enum class State {
	START,
	RECV,
	SEND,
	END,
};

std::ostream &operator<<(std::ostream &os, const State &self) {
	return os << static_cast<std::underlying_type<State>::type>(self);
}

int
main(int argc, char *argv[])
{
	std::minstd_rand rand;
	rand.seed(getpid());

	if (argc < 3) {
		std::cerr << "Not enough args: "  << argv[0] << " <host> <port>" << std::endl;
		exit(EXIT_FAILURE);
	}

	char *host = argv[1], *port = argv[2];

	Net net;
	if (!net.init(host, port)) {
		std::cerr << "Failed to initialise network" << std::endl;
		exit(EXIT_FAILURE);
	}

	State state = State::START;
	std::unique_ptr<Board> board;

	/* initialised to satisfy GCC's linter and sanitiser */
	enum hex_player player = HEX_PLAYER_BLACK;
	enum hex_player opponent = HEX_PLAYER_WHITE;
	enum hex_player winner = HEX_PLAYER_BLACK;

	// game parameters (unused)
	u32 game_secs, thread_limit, mem_limit_mib;
	(void) game_secs; (void) thread_limit; (void) mem_limit_mib;

	bool game_over = false, first_round = true;
	while (!game_over) {
		switch (state) {
		case State::START: {
			std::vector<enum hex_msg_type> expected_msg_types = {HEX_MSG_START};

			struct hex_msg msg;
			if (!net.recv_msg(msg, expected_msg_types)) {
				std::cerr << "Failed to receive message from hex server" << std::endl;
				exit(EXIT_FAILURE);
			}

			player = static_cast<enum hex_player>(msg.data.start.player);
			opponent = hexopponent(player);
			game_secs = msg.data.start.game_secs;
			thread_limit = msg.data.start.thread_limit;
			mem_limit_mib = msg.data.start.mem_limit_mib;

			u32 board_size = msg.data.start.board_size;

			board = std::make_unique<Board>(board_size, rand);

			std::cout << "[" << hexplayerstr(player) << "] Starting game: "
				  << board_size << "x" << board_size << ", "
				  << game_secs << "secs" << std::endl;

			switch (player) {
			case HEX_PLAYER_BLACK: state = State::SEND; break;
			case HEX_PLAYER_WHITE: state = State::RECV; break;
			}
		} break;

		case State::RECV: {
			std::vector<enum hex_msg_type> expected_msg_types = {HEX_MSG_MOVE, HEX_MSG_SWAP, HEX_MSG_END};

			struct hex_msg msg;
			if (!net.recv_msg(msg, expected_msg_types)) {
				std::cerr << "Failed to receive message from hex server" << std::endl;
				exit(EXIT_FAILURE);
			}

			switch (msg.type) {
			case HEX_MSG_MOVE:
				board->play(opponent, msg.data.move.board_x, msg.data.move.board_y);

				if (first_round && rand() % 2) {
					board->swap(rand);

					msg.type = HEX_MSG_SWAP;
					if (!net.send_msg(msg)) {
						std::cerr << "Failed to send message to hex server" << std::endl;
						exit(EXIT_FAILURE);
					}

					state = State::RECV;
				} else {
					state = State::SEND;
				}
				break;

			case HEX_MSG_SWAP:
				board->swap(rand);
				state = State::SEND;
				break;

			case HEX_MSG_END:
				winner = static_cast<enum hex_player>(msg.data.end.winner);
				state = State::END;
				break;
			}

			first_round = false;
		} break;

		case State::SEND: {
			struct hex_msg msg;
			msg.type = HEX_MSG_MOVE;

			Move move;
			if (!board->next(move)) {
				std::cerr << "Failed to generate next board move" << std::endl;
				exit(EXIT_FAILURE);
			}

			board->play(player, move.x, move.y);

			msg.data.move.board_x = move.x;
			msg.data.move.board_y = move.y;

			if (!net.send_msg(msg)) {
				std::cerr << "Failed to send message to hex server" << std::endl;
				exit(EXIT_FAILURE);
			}

			state = State::RECV;
			first_round = false;
		} break;

		case State::END: {
			std::cout << "[" << hexplayerstr(player) << "] Player " << hexplayerstr(winner) << " has won the game" << std::endl;
			game_over = true;
		} break;

		default:
			std::cerr << "Unknown game state: " << state << std::endl;
			exit(EXIT_FAILURE);
			break;
		}
	}

	exit(EXIT_SUCCESS);
}
